File: dsa-problems/Arrays/solutions/java/ARR-006-zero-slide-limit.java
import java.util.*;

class Solution {
    public int[] zeroSlideWithLimit(int[] arr, int m) {
        int n = arr.length;
        int writeIdx = 0;

        for (int readIdx = 0; readIdx < n; readIdx++) {
            if (arr[readIdx] != 0) {
                // If needs to move (i.e., there are zeros behind/writeIdx is behind)
                if (readIdx != writeIdx) {
                    if (m <= 0) break; // Limit reached

                    // Swap
                    int temp = arr[writeIdx];
                    arr[writeIdx] = arr[readIdx];
                    arr[readIdx] = temp;

                    m--;
                }
                writeIdx++;
            }
        }
        return arr;
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt()) return;
        int n = sc.nextInt();
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) arr[i] = sc.nextInt();
        int m = sc.nextInt();

        Solution solution = new Solution();
        int[] result = solution.zeroSlideWithLimit(arr, m);

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; i++) {
            sb.append(result[i]).append(i == n - 1 ? "" : " ");
        }
        System.out.println(sb);
        sc.close();
    }
}

File: dsa-problems/Arrays/solutions/python/ARR-006-zero-slide-limit.py
import sys

def zero_slide_with_limit(arr: list[int], m: int) -> list[int]:
    """
    Move zeros to end with swap limit.
    """
    n = len(arr)
    write_idx = 0

    for read_idx in range(n):
        if arr[read_idx] != 0:
            if read_idx != write_idx:
                if m <= 0:
                    break
                # Swap
                arr[write_idx], arr[read_idx] = arr[read_idx], arr[write_idx]
                m -= 1
            write_idx += 1

    return arr

def main():
    n = int(input())
    arr = list(map(int, input().split()))
    m = int(input())

    result = zero_slide_with_limit(arr, m)
    print(" ".join(map(str, result)))

if __name__ == "__main__":
    main()

